src/realm/object-store/c_api/socket_provider.cpp

```diff
@@ -0,0 +1,255 @@
+    }
+
+    bool websocket_binary_message_received(util::Span<const char> data) final
+    {
+        return m_observer.websocket_binary_message_received(data);
+    }
+
+    bool websocket_closed_handler(bool was_clean, Status status) final
+    {
+        return m_observer.websocket_closed_handler(was_clean, status);
+    }
+
+private:
+    sync::WebSocketObserver& m_observer;
+};
+
+struct CAPISyncSocketProvider : sync::SyncSocketProvider {
+    realm_userdata_t m_userdata = nullptr;
+    realm_free_userdata_func_t m_free = nullptr;
+    realm_sync_socket_post_func_t m_post = nullptr;
+    realm_sync_socket_create_timer_func_t m_timer_create = nullptr;
+    realm_sync_socket_timer_canceled_func_t m_timer_cancel = nullptr;
+    realm_sync_socket_timer_free_func_t m_timer_free = nullptr;
+    realm_sync_socket_connect_func_t m_websocket_connect = nullptr;
+    realm_sync_socket_websocket_async_write_func_t m_websocket_async_write = nullptr;
+    realm_sync_socket_websocket_free_func_t m_websocket_free = nullptr;
+
+    CAPISyncSocketProvider() = default;
+    CAPISyncSocketProvider(CAPISyncSocketProvider&& other)
+        : m_userdata(std::exchange(other.m_userdata, nullptr))
+        , m_free(std::exchange(other.m_free, nullptr))
+        , m_post(std::exchange(other.m_post, nullptr))
+        , m_timer_create(std::exchange(other.m_timer_create, nullptr))
+        , m_timer_cancel(std::exchange(other.m_timer_cancel, nullptr))
+        , m_timer_free(std::exchange(other.m_timer_free, nullptr))
+        , m_websocket_connect(std::exchange(other.m_websocket_connect, nullptr))
+        , m_websocket_async_write(std::exchange(other.m_websocket_async_write, nullptr))
+        , m_websocket_free(std::exchange(other.m_websocket_free, nullptr))
+    {
+        REALM_ASSERT(m_free);
+        REALM_ASSERT(m_post);
+        REALM_ASSERT(m_timer_create);
+        REALM_ASSERT(m_timer_cancel);
+        REALM_ASSERT(m_timer_free);
+        REALM_ASSERT(m_websocket_connect);
+        REALM_ASSERT(m_websocket_async_write);
+        REALM_ASSERT(m_websocket_free);
+    }
+
+    ~CAPISyncSocketProvider()
+    {
+        m_free(m_userdata);
+    }
+
+    std::unique_ptr<sync::WebSocketInterface> connect(sync::WebSocketObserver* observer,
+                                                      sync::WebSocketEndpoint&& endpoint) final
+    {
+        auto capi_observer = std::make_shared<CAPIWebSocketObserver>(*observer);
+        return std::make_unique<CAPIWebSocket>(m_userdata, m_websocket_connect, m_websocket_async_write,
+                                               m_websocket_free, new realm_websocket_observer_t(capi_observer),
+                                               std::move(endpoint));
+    }
+
+    void post(FunctionHandler&& handler) final
+    {
+        auto shared_handler = std::make_shared<FunctionHandler>(std::move(handler));
+        m_post(m_userdata, new realm_sync_socket_callback_t(std::move(shared_handler)));
```

Comment:
where does the pointer we allocate here get free'd? does the post function call realm_release() on it?

Reply:
`realm_sync_socket_callback_complete` should free the pointer. I will update it.

Reply:
I had a look at the pointers that are allocated by this api. `realm_sync_socket_callback_t*` is free'd by `realm_sync_socket_callback_complete`. `realm_websocket_observer_t*` is now free'd when the c api socket is free'd.